IQSS/dataverse
# Intellij IDEA webapp hotreload

**What this PR does / why we need it**:

This is related to https://github.com/IQSS/dataverse/pull/9959.

When running Dataverse in Docker we still want to be able to just edit things under `src/main/webapp` using IntelliJ IDEA and then just reload the web page to see the changes. To do this in this PR:

1. Mapped Payara `/opt/payara/appserver/glassfish/domains/domain1/applications` folder to `./docker-dev-volumes/glassfish/applications`
2. Added `watchers.xml` File watcher configuration, which can be imported into  IDEA to ...
3. ... run `cpwebapp.sh` to copy changed files under src/main/webapp to `./docker-dev-volumes/glassfish/applications/dataverse-{current version}`

**Which issue(s) this PR closes**:

None. 

**Special notes for your reviewer**:

You need IntelliJ IDEA Ultimate edition for File watchers support.

**Suggestions on how to test this**:

1. Run Dataverse in Docker https://guides.dataverse.org/en/latest/container/index.html. 
2. In IntelliJ IDEA, import `scripts/intellij/watchers.xml` into File watchers (https://www.jetbrains.com/help/idea/settings-tools-file-watchers.html).
3. Setup Payara remote run config and deploy Dataverse in debug mode (https://dataverse-guide--9959.org.readthedocs.build/en/9959/container/dev-usage.html)
4. Open Dataverse in browser.
5. Edit and save any XHTML/CSS/etc. file under `src/main/webapp`.
6. Reload browser to see changes to the UI.

See this here in action: https://harvard.zoom.us/rec/share/EUOqzaq4NzsH4N8xX7kg9II8w292MEB6eqbyzPMlZOACsErKV-Z20hHBj1HerFH2.bGFh04EdU1Qnu9ic

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No.

**Is there a release notes update needed for this change?**:

Yes.

**Additional documentation**:

The documentation for this is practically what is written in `Suggestions on how to test this`.
